graphql_compiler/cost_estimation/analysis.py

```diff
@@ -0,0 +1,256 @@
+    will have similar sizes. This reasoning is local: if a filter in a different location is
+    correlated with the values on this field, the generated pages might turn out to have
+    wildly different sizes. This problem is somewhat unavoidable.
+
+    Args:
+        schema_info: QueryPlanningSchemaInfo
+        field_value_intervals: see get_field_value_intervals
+        distinct_result_set_estimates: see get_distinct_result_set_estimates
+        query_metadata: info on locations, inputs, outputs, and tags in the query
+        parameters: the query parameters
+
+    Returns:
+        the pagination capacity of each PropertyPath
+    """
+    pagination_capacities = {}
+    for location, location_info in query_metadata.registered_locations:
+        vertex_type_name = location_info.type.name
+
+        for field_name, _ in location_info.type.fields.items():
+            property_path = PropertyPath(location.query_path, field_name)
+            if not is_meta_field(field_name):
+                if is_uuid4_type(schema_info, vertex_type_name, field_name):
+                    pagination_capacities[property_path] = distinct_result_set_estimates[
+                        location.query_path
+                    ]
+                elif field_supports_range_reasoning(schema_info, vertex_type_name, field_name):
+                    field_value_interval = field_value_intervals.get(
+                        property_path, Interval(None, None)
+                    )
+                    quantiles = schema_info.statistics.get_field_quantiles(
+                        vertex_type_name, field_name
+                    )
+                    if quantiles is not None:
+
+                        # Since we can't be sure the minimum observed value is the
+                        # actual minimum value, we treat values less than it as part
+                        # of the first chunk. Similarly, we treat values greater than the known
+                        # maximum as part of the last chunk. That's why we drop the minimum and
+                        # maximum observed values from the quantile list.
+                        proper_quantiles = quantiles[1:-1]
+
+                        # Get the relevant quantiles (ones inside the field_value_interval)
+                        min_quantile = 0
+                        max_quantile = len(proper_quantiles)
+                        if field_value_interval.lower_bound is not None:
+                            min_quantile = bisect.bisect_left(
+                                proper_quantiles, field_value_interval.lower_bound
+                            )
+                        if field_value_interval.upper_bound is not None:
+                            max_quantile = bisect.bisect_left(
+                                proper_quantiles, field_value_interval.upper_bound
+                            )
+                        relevant_quantiles = proper_quantiles[min_quantile:max_quantile]
```

Comment:
This code still doesn't make sense to me based on the comment above. Based on your comment, above we should make 4 pages out of the [1,2,3]  quantiles:  [[-inf, 1), [1,2), [2,3), [3, inf]]. However, assuming no filter bounds and a distinct result estimate of infinity, we would get a pagination capacity of 2 no?